src/main/java/org/jenkinsci/plugins/workflow/graphanalysis/ForkScanner.java

```diff
@@ -0,0 +1,444 @@
+import java.util.HashMap;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.Set;
+
+/**
+ * Scanner that will scan down all forks when we hit parallel blocks before continuing, but generally runs in linear order
+ * <p/>Think of it as the opposite of {@link DepthFirstScanner}.
+ *
+ * <p/>This is a fairly efficient way to visit all FlowNodes, and provides three useful guarantees:
+ * <ul>
+ *   <li>Every FlowNode is visited, and visited EXACTLY ONCE (not true for LinearScanner)</li>
+ *   <li>All parallel branches are visited before we move past the parallel block (not true for DepthFirstScanner)</li>
+ *   <li>For EVERY block, the BlockEndNode is visited before the BlockStartNode (not true for DepthFirstScanner, with parallels)</li>
+ * </ul>
+ *
+ * <p/>The big advantages of this approach:
+ * <ul>
+ *     <li>Blocks are visited in the order they end (no backtracking) - helps with working a block at a time</li>
+ *     <li>Points are visited in linear order within a block (easy to use for analysis)</li>
+ *     <li>Minimal state information needed</li>
+ *     <li>Branch information is available for use here</li>
+ * </ul>
+ *
+ * @author <samvanoort@gmail.com>Sam Van Oort</samvanoort@gmail.com>
+ */
+public class ForkScanner extends AbstractFlowScanner {
+
+    // Last element in stack is end of myCurrent parallel start, first is myCurrent start
+    ArrayDeque<ParallelBlockStart> parallelBlockStartStack = new ArrayDeque<ParallelBlockStart>();
+
+    /** FlowNode that will terminate the myCurrent parallel block */
+    FlowNode currentParallelStartNode = null;
+
+    ParallelBlockStart currentParallelStart = null;
+
+    private boolean walkingFromFinish = false;
+
+    @Override
+    protected void reset() {
+        parallelBlockStartStack.clear();
+        currentParallelStart = null;
+        currentParallelStartNode = null;
+        myCurrent = null;
+        myNext = null;
+    }
+
+    /** If true, we are walking from the flow end node and have a complete view of the flow */
+    public boolean isWalkingFromFinish() {
+        return walkingFromFinish;
+    }
+
+    /** Tracks state for parallel blocks, so we can ensure all are visited and know the branch starting point */
+    protected static class ParallelBlockStart {
+        protected BlockStartNode forkStart; // This is the node with child branches
+        protected int remainingBranches;
+        protected int totalBranches;
+        protected ArrayDeque<FlowNode> unvisited = new ArrayDeque<FlowNode>();  // Remaining branches of this that we have have not visited yet
+
+        protected ParallelBlockStart(BlockStartNode forkStart, int branchCount) {
+            this.forkStart = forkStart;
+            this.remainingBranches = branchCount;
+        }
+
+        /** Strictly for internal use in the least common ancestor problem */
+        ParallelBlockStart() {}
+    }
+
+    interface FlowPiece {  // Mostly a marker
```

Comment:
No, piece == one or more nodes.  I just needed a superinterface that they could both implement (for container use, so I wasn't just passing around/casting arbitrary objects).